### PR TITLE
Route resource alerts via Email tag with User fallback

### PR DESCRIPTION
**Pull request by pragya811**
## Type of change
Note: Fill **x** in []
- [ ] bug
- [x] enhancement
- [ ] documentation
- [ ] dependencies

## Description
Lets a resource owner self-service redirect cleanup/zombie alerts to a team distribution list by setting the Email tag, instead of always going to a single individual via the User tag.

- Add resolve_alert_recipient() / is_valid_alert_email() helper: prefers a resource's Email tag (validated for format, header-injection characters, and an allowed domain list), falls back to the User tag when Email is unset, invalid, or 'NA' - no behavior change for existing resources without an Email tag.
- Wire the helper into ec2_stop.py and zombie_cluster_common_methods.py, the two live per-resource alert paths. LDAP/manager-CC lookups stay keyed on the User tag, never on the resolved recipient.
- Add ALLOWED_EMAIL_DOMAINS env var (default '@redhat.com') so the domain check is configurable for other deployments.
- Fix a pre-existing bug in Postfix.prettify_to/prettify_cc where the "is this already a full address" check used substring matching instead of a proper domain suffix match, letting a value like victim@redhat.com.evil.com bypass the redhat.com-only assumption.

### CODE REVIEW

File: cloud_governance/policy/aws/ec2_stop.py [hunk 1]
```diff
@@ -128,8 +129,14 @@ def __trigger_mail(self, tags: list, stopped_time: str, resource_id: str, days:
             user, instance_name = self._get_tag_name_from_tags(tags=tags,
                                                                tag_name='User'), self._get_tag_name_from_tags(tags=tags,
                                                                                                               tag_name='Name')
-            to = user if user not in special_user_mails else special_user_mails[user]
-            ldap_data = self._ldap.get_user_details(user_name=to)
+            if user in special_user_mails:
+                to = special_user_mails[user]
+                ldap_lookup_user = to
+            else:
+                email = self._get_tag_name_from_tags(tags=tags, tag_name='Email')
```
Comment (ebattat) [inline cloud_governance/policy/aws/ec2_stop.py]: tag_name='Email'
What will be if the user will add 'email'/'EMAIL', will it work ?

File: cloud_governance/policy/policy_operations/aws/zombie_cluster/zombie_cluster_common_methods.py [hunk 2]
```diff
@@ -224,10 +225,16 @@ def trigger_mail(self, tags: list, resource_id: str, days: int, resources: list,
                 tags=tags, tag_name='Name')
             if not resource_name:
                 resource_name = self.get_tag_name_from_tags(tags=tags, tag_name='cg-Name')
-            to = user if user not in special_user_mails else special_user_mails[user]
-            ldap_data = self._ldap.get_user_details(user_name=to)
+            if user in special_user_mails:
+                to = special_user_mails[user]
+                ldap_lookup_user = to
+            else:
+                email = self.get_tag_name_from_tags(tags=tags, tag_name='Email')
```
Comment (ebattat) [inline cloud_governance/policy/policy_operations/aws/zombie_cluster/zombie_cluster_common_methods.py]: Same question as above ?

File: cloud_governance/policy/aws/ec2_stop.py (hunk 1 above)
Comment (pragya811) [reply cloud_governance/policy/aws/ec2_stop.py]: 'Email' tag is added by the automation which the user can then modify. However I have added a case insensitive matching which should cover the above gap.

Comment (ebattat) [reply cloud_governance/policy/aws/ec2_stop.py]: So by default it is empty, right ?

File: cloud_governance/common/mails/alert_recipient.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    """
+    This method resolves the alert recipient for a resource, preferring the Email tag
+    (self-service group/team routing) and falling back to the User tag - matching today's
+    behavior when Email is unset or invalid.
+    @param email_tag_value:
+    @param user_tag_value:
+    @return:
+    """
+    if is_valid_alert_email(email_tag_value):
+        return email_tag_value.strip()
```
Comment (ebattat) [inline cloud_governance/common/mails/alert_recipient.py]: **Issue #1: Missing validation for empty User tag fallback**

When the Email tag is invalid/missing, the code falls back to `user_tag_value` without validating it. If both tags are empty or 'NA', this could pass an invalid recipient downstream causing runtime errors.

**Recommendation:**
```python
def resolve_alert_recipient(email_tag_value: str, user_tag_value: str) -> str:
    if is_valid_alert_email(email_tag_value):
        return email_tag_value.strip()
    # Validate user_tag_value is not empty/NA before returning
    user = (user_tag_value or ').strip()
    if user and user.upper() != 'NA':
        return user
    return ''  # Return empty string if both tags are invalid
```

This prevents silent failures when both Email and User tags are missing/invalid.

**Review by ebattat — COMMENTED** [review]
Found two issues that I recommend addressing before merge:

1. **Empty User tag validation** - prevents runtime errors when both tags are missing/invalid
2. **Expanded header injection checks** - defense-in-depth improvement

See inline comments for details.

File: cloud_governance/common/mails/alert_recipient.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    """
+    This method validates a resource's Email tag value before it is used as an alert
+    recipient: well-formed, free of header-injection characters, and on an allowed domain.
+    @param email:
+    @return:
+    """
+    if not email:
+        return False
+    email = email.strip()
+    if not email or email.upper() == 'NA' or '\n' in email or '\r' in email:
```
Comment (ebattat) [inline cloud_governance/common/mails/alert_recipient.py]: **Issue #2: Incomplete header injection protection**

The validation checks for `\n` and `\r` but not other potential header injection vectors like horizontal tabs (`\t`) or null bytes (`\x00`).

**Recommendation:** Expand the check for defense-in-depth:
```python
if not email or email.upper() == 'NA' or any(c in email for c in '\n\r\t\x00'):
    return False
```

This provides more comprehensive protection against header injection attacks.

File: cloud_governance/policy/aws/ec2_stop.py (hunk 1 above)
Comment (pragya811) [reply cloud_governance/policy/aws/ec2_stop.py]: By default it has the value derived from 'User', same as how it's functioning today. But if the user edits it to a group email, the alerts will go to that resource.

File: tests/unittest/cloud_governance/common/postfix/test_alert_recipient.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    tags = [{'Key': 'EMAIL', 'Value': 'team-dl@redhat.com'}]
+    assert get_email_tag_value(tags=tags) == 'team-dl@redhat.com'
+
+
+def test_get_email_tag_value_returns_empty_when_absent():
+    """
+    This method tests an empty string is returned when no Email tag exists
+    """
+    tags = [{'Key': 'User', 'Value': 'jdoe'}]
+    assert get_email_tag_value(tags=tags) == ''
```
Comment (ebattat) [inline tests/unittest/cloud_governance/common/postfix/test_alert_recipient.py]: **Suggestion: Add test for both tags missing/invalid**

Consider adding a test case for when both Email and User tags are empty or 'NA':

```python
def test_resolve_alert_recipient_returns_empty_when_both_tags_invalid():
    """
    This method tests that an empty string is returned when both Email and User tags are invalid
    """
    _set_allowed_domains(['@redhat.com'])
    assert resolve_alert_recipient(email_tag_value='NA', user_tag_value='NA') == ''
    assert resolve_alert_recipient(email_tag_value=''', user_tag_value=''') == ''
```

This ensures the edge case is handled gracefully (related to Issue #1).

File: cloud_governance/common/mails/alert_recipient.py (hunk 3 above)
Comment (pragya811) [reply cloud_governance/common/mails/alert_recipient.py]: resolve_alert_recipient() now validates user_tag_value too — an empty/NA User tag returns '' instead of silently passing through a placeholder.

File: cloud_governance/common/mails/alert_recipient.py (hunk 4 above)
Comment (pragya811) [reply cloud_governance/common/mails/alert_recipient.py]: The specific example is already handled by the regex, but checking confirmed a genuine one: a null byte (\x00) is not whitespace, so it slips past the regex entirely. Added a broader check (ord(character) < 32, i.e. any C0 control character) rather than their enumerated list — this also catches other non-whitespace control characters they didn't name (e.g. \x01, escape)

File: tests/unittest/cloud_governance/common/postfix/test_alert_recipient.py (hunk 5 above)
Comment (pragya811) [reply tests/unittest/cloud_governance/common/postfix/test_alert_recipient.py]: Added

**Review by ebattat — APPROVED** [review]
/approve
